Intents(in) for LFRic objects (for #712)

TeranIvy opened the pull request:
This PR is for issue #712 and it modifies the PSy-layer intents to `in` for all objects whose data are updated by their proxies, that is Field (including Field vectors), Operator and Columnwise operator derived types. The intent of Scalars is still defined by their access metadata as they are actual data and PSyclone only allows `GH_READ` for the user-defined scalars anyway.

Tests with compilation pass. LFRic was successfully tested in [MOSRS ticket #2097](https://code.metoffice.gov.uk/trac/lfric/ticket/2097) (see [comment:17](https://code.metoffice.gov.uk/trac/lfric/ticket/2097#comment:17) and [comment:19](https://code.metoffice.gov.uk/trac/lfric/ticket/2097#comment:19)).

There are not many changes in `dynamo0p3.py` source but they touch many tests. `dynamo0p3.rst` document is also updated with the description of argument intents in PSy and Kernel layer and the explanation of intents in built-ins.

LFRic examples (`examples/lfric`) are also updated:
- Appropriate intents in algorithms and PSyKAl-lite code (following the LFRic code as much as possible),
- Correct copyrights in modified code,
- Removal of constructors from LFRic kernels (changed in LFRic code last year).

Ready for first review from either @arporter or @rupertford.

arporter commented on examples/lfric/eg4/solver_mod.x90:
```diff
@@ -107,8 +107,9 @@ subroutine jacobi_solver_algorithm(lhs, rhs, mm, mesh, n_iter)
   integer(i_def),      intent(in)    :: n_iter
-  type(field_type),    intent(inout) :: lhs, rhs
-  type(operator_type), intent(inout) :: mm
+  type(field_type),    intent(inout) :: lhs
```
Please could you explain why this is `inout` and not `in`?

TeranIvy replied:
Because of the LFRic code equivalents where particular field intents had to be left as ``inout`` for successful compilation.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2669,26 +2669,22 @@ class DynFields(DynCollection):
-                                       entity_decls=fld_args[intent],
-                                       intent=fort_intent))
+
+        # Add the Invoke subroutine argument declarations for fields
```
Please extend this comment to explain why we can safely declare all field objects to have intent in.

TeranIvy replied:
For consistency with similar comments I left this one as as and updated the `_invoke_declarations` docstring along the lines of the one for operators.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2994,28 +2990,22 @@ def _stub_declarations(self, parent):
         '''
         Declare all LMA-related quantities in a PSy-layer routine.
+        Note: PSy layer in LFRic does not modify the LMA operator objects.
+        Hence, their Fortran intents are always "in" (the data pointed to
```
s/the data pointed to by the operators are updated in kernels)/the data updated in the kernels is only *pointed* to from the operator object and is thus not a part of the object)/

TeranIvy replied:
Thanks for the suggestion, done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3134,6 +3124,9 @@ def _invoke_declarations(self, parent):
         Generate the necessary PSy-layer declarations for all column-wise
         operators and their associated parameters.
+        Note: PSy layer in LFRic does not modify the CMA operator objects.
+        Hence, their Fortran intents are always "in" (the data pointed to
```
Please update this docstring along the lines of the one for LMA quantities.

TeranIvy replied:
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -100,11 +110,20 @@ Please see the :ref:`algorithm-layer` section for a description of the
+Scalar
+++++++
+
+In Dynamo0.3 API a scalar is a single value variable that can be
```
Actually I think they are my errors as I added this section 😃 Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1660,6 +1704,27 @@ arguments to inter-grid kernels are as follows:
+As described :ref:`above <dynamo0.3-psy-arg-intents>`, LFRic kernels read
+and/or update the data pointed to by the objects such as
+:ref:`fields <dynamo0.3-field>` or :ref:`operators <dynamo0.3-operator>`.
+These data are passed to the kernels as :ref:`subroutine arguments
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1660,6 +1704,27 @@ arguments to inter-grid kernels are as follows:
+################
+
+As described :ref:`above <dynamo0.3-psy-arg-intents>`, LFRic kernels read
+and/or update the data pointed to by the objects such as
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1689,9 +1754,14 @@ The Built-ins supported for the Dynamo0.3 API are listed in the related
 the calculation performed by each Built-in is described using Fortran array
 syntax; this does not necessarily reflect the actual implementation of the
-Built-in (*e.g.* it could be implemented by PSyclone generating a call to an
+Built-in (*e.g.*\ it could be implemented by PSyclone generating a call to an
```
Yep, removed.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1689,9 +1754,14 @@ The Built-ins supported for the Dynamo0.3 API are listed in the related
 optimised maths library).
 
+As described in the PSy-layer :ref:`Argument Intents
+<dynamo0.3-psy-arg-intents>` section, LFRic :ref:`field <dynamo0.3-field>`
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1689,9 +1754,14 @@ The Built-ins supported for the Dynamo0.3 API are listed in the related
+As described in the PSy-layer :ref:`Argument Intents
+<dynamo0.3-psy-arg-intents>` section, LFRic :ref:`field <dynamo0.3-field>`
+objects' Fortran intent is always ``in``. The field or scalar whose data is
+modified by a Built-in is marked in **bold**.
```
Done.

TeranIvy commented on examples/lfric/eg2/v3_solver_kernel_mod.F90:
```diff
@@ -34,54 +68,42 @@ module v3_solver_kernel_mod
   integer, intent(in) :: map(ndf)
-!  real(kind=dp), intent(in), dimension(ndf,ngp,ngp,ngp,1) :: v3_basis  
-  real(kind=dp), intent(in), dimension(1,ndf,ngp_h,ngp_v) :: v3_basis  
+!  real(kind=dp), intent(in), dimension(ndf,ngp,ngp,ngp,1) :: v3_basis
```
Done.

TeranIvy commented on examples/lfric/eg2/dynamo_algorithm_mod.F90:
```diff
@@ -35,18 +68,18 @@ subroutine dynamo_algorithm( pressure_density, rhs, &
+    type( field_type ), intent( in )    :: rhs
+    type( field_type ), intent( in )    :: flux_velocity
+    type( field_type ), intent( in )    :: rhs_v2
     type( field_type ), intent( inout ) :: circulation
```
I don't think so. As explained in the PR description, I looked into LFRic equivalents of PSyclone algorithms to determine intents. According to [LFRic ticket 2097, comment 19](https://code.metoffice.gov.uk/trac/lfric/ticket/2097#comment:19) some field intents in algorithms had to be left ``inout`` for successful compilation.

TeranIvy commented on examples/lfric/eg2/solver_mod.F90:
```diff
@@ -24,20 +56,20 @@ module solver_mod
     implicit none
-    type(field_type), intent(inout)    :: lhs
-    type(field_type), intent(inout)    :: rhs
+    type(field_type), intent(inout) :: lhs
```
Again, I looked into LFRic code equivalents and some field intents had to be left as ``inout`` for successful compilation.

TeranIvy commented on examples/lfric/eg3/solver_mod.x90:
```diff
@@ -426,18 +423,19 @@ contains
     integer(kind=i_def), intent(in)    :: n_iter
-    type(field_type),    intent(inout) :: lhs, rhs
-    type(operator_type), intent(inout) :: mm
+    type(field_type),    intent(inout) :: lhs
```
Because of the LFRic code equivalents where particular field intents had to be left as ``inout`` for successful compilation.

TeranIvy commented on examples/lfric/eg2/dynamo_algorithm_mod.F90:
```diff
@@ -35,18 +68,18 @@ subroutine dynamo_algorithm( pressure_density, rhs, &
+
     call invoke ( v3_rhs_kernel_type(rhs),                            &
                   v3_solver_kernel_type(pressure_density,rhs) )
     !RF call invoke_rhs_v3( rhs )
```
I removed both lines as they are calls to PSyKAl-lite equivalents of the invoke above. I think they were in the original LFRic code before it was used as an example in PSyclone.

TeranIvy commented:
Addressed review comments and brought the branch up to the head of master. Also
- Removed constructors from kernels in `examples/lfric/code` (forgotten previously),
- Cleaned up code,
- Clarified intent for kernel arguments with `GH_WRITE` access (can be `out` or `inout`) in `dynamo0p3.rst`.

Ready for second review from @arporter.

arporter commented:
All requested changes have been made.
All changes are covered.
All tests pass, incl. with compilation. Ditto for examples.
